### PR TITLE
Introduce Pin structs

### PR DESCRIPTION
**Pull request by lack**
This replaces all use of map[string]interface{} with well-defined
structs.

Closes issue #4

### CODE REVIEW

**Comment by lack** [comment]
Here's my first attempt.  I've done manual tests for tang, sss and tpm2 (I don't have a yubikey), showing they can properly interact with latchset/clevis.

I hope I didn't go too overboard.  It adds Pin structs for all the supported pin types, but also a "ToConfig" operation to convert a Pin to the associated Config which can be used to encrypt something new with the same pin. And some minor cleanup like `go lint` fixes and moving some of the encryption code a bit to make it more user-friendly.

This doesn't have any unit test updates yet.  If you approve of the general approach I'll add them.

Side-note: tpm2 looks broken (cannot Decrypt a tpm2 payload encrypted by latchset/clevis), not just in this code but also in main.  I'll be raising an issue about that too, soon.

File: yubikey.go [hunk 1]
```diff
@@ -18,35 +18,65 @@ import (
...
 
-func DecryptYubikey(msg *jwe.Message, clevisNode map[string]interface{}) ([]byte, error) {
-	yubikeyNode, ok := clevisNode["yubikey"].(map[string]interface{})
-	if !ok {
-		return nil, fmt.Errorf("clevis.go/yubikey: cannot parse provided token, node 'clevis.yubikey'")
-	}
+// YubikeyPin represents the data yubikey needs to perform decryption
+type YubikeyPin struct {
+	Type      string     `json:"type"`
+	Challenge string     `json:"chalelenge"`
```
Comment (anatol) [inline yubikey.go:24]: The json field name needs to be `challenge`, otherwise tests fail.

File: sss.go [hunk 2]
```diff
@@ -11,19 +11,93 @@ import (
...
+func (p SssPin) ToConfig() (SssConfig, error) {
+	c := SssConfig{
+		Threshold: p.Threshold,
+		Pins:      make(map[string][]json.RawMessage),
 	}
+	jweErrors := make([]error, 0)
+	for _, jwe := range p.Jwe {
+		_, pin, err := Parse([]byte(jwe))
+		if err != nil {
+			jweErrors = append(jweErrors, err)
```
Comment (anatol) [inline sss.go:39]: This kind of error accumulation look not very Go idiomatic. Would it make a sense return the first error and use `return err` everywhere in this function?

File: sss.go [hunk 3]
```diff
@@ -11,19 +11,93 @@ import (
...
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+		case "sss":
+			cfg, err := pin.Sss.ToConfig()
+			if err != nil {
+				jweErrors = append(jweErrors, err)
+				continue
+			}
+			c.Pins["sss"], err = appendConfig(c.Pins["sss"], cfg)
```
Comment (anatol) [inline sss.go:71]: `appendConfig` repeats in the all `case` branches. For me it looks like it can be moved outside of `switch` statement.

And `appendConfig` can be inlined as it is a trivial function (i.e. just call marshal and append directly without introducing a new function).

File: tang.go [hunk 4]
```diff
@@ -20,19 +20,44 @@ import (
 	"github.com/lestrrat-go/jwx/jws"
 )
 
-// DecryptTang decrypts a jwe message bound with Tang clevis pin
-func DecryptTang(msg *jwe.Message, clevisNode map[string]interface{}) ([]byte, error) {
-	tangNode, ok := clevisNode["tang"].(map[string]interface{})
-	if !ok {
-		return nil, fmt.Errorf("cannot parse provided token, node 'clevis.tang'")
+// TangPin  represents the data tang needs to perform decryption
+type TangPin struct {
```
Comment (anatol) [inline tang.go:24]: Some structs (like `*Pin` and `*Config`) are uppercase and they become public. Do we really need to expose it to users? If not then it should be private.

**Review by anatol — CHANGES_REQUESTED** [review]
Fantastic change. Thank you. With the Yubikey code fix all the tests pass for me.

wrt unit tests - I would really appreciate if you add it. If you see use-cases or corner cases that are not covered by tests - please add it. The tests will pay back later.

File: yubikey.go (hunk 1 above)
Comment (lack) [reply yubikey.go:24]: Good catch!  Will fix.

File: sss.go (hunk 3 above)
Comment (lack) [reply sss.go:71]: But since every 'cfg' is a different type, I'm not sure how to do that....  I did have a version of the code that used an interface so I could have common code for any `interface{ ToConfig() (config, error) }`, but the fact that this changes `ToConfig` to have to return an interface as well, which then has other implications elsewhere in the code...

File: sss.go [hunk 5]
```diff
@@ -11,19 +11,93 @@ import (
...
-	if !ok {
-		return nil, fmt.Errorf("clevis.go/sss: cannot parse provided token, node 'clevis.sss.p'")
+// ToConfig converts a given SssPin into the corresponding SssConfig which can be used for encryption
+func (p SssPin) ToConfig() (SssConfig, error) {
+	c := SssConfig{
+		Threshold: p.Threshold,
+		Pins:      make(map[string][]json.RawMessage),
 	}
+	jweErrors := make([]error, 0)
+	for _, jwe := range p.Jwe {
```
Comment (anatol) [inline sss.go:36]: Intellij warns: `Variable `jwe` collides with imported package name`.

File: tang.go (hunk 4 above)
Comment (lack) [reply tang.go:24]: I was thinking they should be public, since I can envision how I'd want to use this library, to interoperate with the `*Pin` and `*Config` structures in code.

File: sss.go (hunk 3 above)
Comment (anatol) [reply sss.go:71]: I was thinking of making it `var cfg interface{}` and putting it before the `switch`.

File: sss.go (hunk 5 above)
Comment (lack) [reply sss.go:36]: Interesting `go lint` didn't complain, but I'll rename that :)

File: tang.go (hunk 4 above)
Comment (lack) [reply tang.go:24]: That said, there may be some of the functions I introduced as public that don't need to be...  I'll take a pass at cleaning that up a bit as I work on the unit tests.

**Comment by lack** [comment]
This latest push addresses all review comments and adds unit tests for the critical sections.

File: sss.go (hunk 2 above)
Comment (lack) [reply sss.go]: Fixed with a more go-like fail-first approach.

File: sss.go (hunk 3 above)
Comment (lack) [reply sss.go:64]: Great idea!  It's much simpler now.

File: tang.go (hunk 4 above)
Comment (lack) [reply tang.go:24]: I have moved a few functions to be non-Public.

**Comment by anatol** [comment]
The change look good. Merging it now.

**Comment by anatol** [comment]
Thank you for your work Jim!

**Comment by lack** [comment]
> Thank you for your work Jim!

Glad to help!
